test/common/event/file_event_impl_test.cc

```diff
@@ -439,6 +439,41 @@ TEST_F(FileEventImplTest, RegisterIfEmulatedEdge) {
   dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
 }
 
+#ifndef WIN32 // This is a test case that clarifies the event behavior.
+// Read and Write events are activated by the kernel before invoke connect() or listen().
+TEST_P(FileEventImplActivateTest, EventTriggeredBeforeConnect) {
+  os_fd_t fd = os_sys_calls_.socket(domain(), SOCK_STREAM, 0).return_value_;
+  ASSERT_TRUE(SOCKET_VALID(fd));
+
+  Api::ApiPtr api = Api::createApiForTest();
+  DispatcherPtr dispatcher(api->allocateDispatcher("test_thread"));
+  ReadyWatcher read_event;
+  ReadyWatcher write_event;
+
+  // Emulate the behavior of the ConnectionImpl constructor which initializes the file event before
+  // connect().
+  Event::FileEventPtr file_event = dispatcher->createFileEvent(
+      fd,
+      [&](uint32_t events) -> void {
+        if (events & FileReadyType::Read) {
+          read_event.ready();
+        }
+
+        if (events & FileReadyType::Write) {
+          write_event.ready();
+        }
+      },
+      Event::PlatformDefaultTriggerType, FileReadyType::Read | FileReadyType::Write);
+
+  EXPECT_CALL(read_event, ready());
+  EXPECT_CALL(write_event, ready());
```

Comment:
It is unclear why a read event is delivered.  My best guess is that the socket is not fully created to the read event is the kernel complaining about an error.

Reply:
Yeah, so we should either prevent the the flow described in this test case.
Or we fix the client connection to be safe even this flow is triggered.